React/Views/ScrollView/RCTScrollView.m

```diff
@@ -185,9 +185,7 @@ - (instancetype)initWithFrame:(CGRect)frame
   if ((self = [super initWithFrame:frame])) {
 #if TARGET_OS_OSX // [TODO(macOS ISS#2323203)
     self.scrollEnabled = YES;
-    self.hasHorizontalScroller = YES;
     self.hasVerticalScroller = YES;
```

Comment:
I think we need a similar affordance for when the ScrollView is `horizontal={true}`.   I tried ScrollViewExample and its horizontal tests have what looks like a similar problem to the `horizontal={false}` but on the other axis:

<img width="774" alt="horizontal" src="https://user-images.githubusercontent.com/30053638/84837345-bad6bf80-afec-11ea-8019-7a742d872591.png">

Does it just need to not call `hasVerticalScroller` here and set it conditionally down on line 754 like it is for `setHasHorizontalScroller` ?

Reply:
Almost:
<img width="752" alt="image" src="https://user-images.githubusercontent.com/4123478/84897177-7fe49280-b0a5-11ea-8fed-6f5f5eb1f1d2.png">
The problem is that we don't regain the area and the view appears clipped. This has always been there but it's more obvious with no vertical scroller. :disappointed:

Reply:
I checked it out and it so close!    From experimentation, it seems like its the `  [[self scrollView] setHasVerticalScroller:!isHorizontal || scrollsVertically];` call on line 757.   In the `horizontal={true}` case, the `scrollsVertically` is true which it should not?   If I simply omit the ` || scrollsVertically` it seems to work

Reply:
I found another problem with both horizontal and vertical.   When vertical, the layout height of the content can be cut off:

<img width="752" alt="vertical" src="https://user-images.githubusercontent.com/30053638/84929134-42234200-b084-11ea-8ccc-d31ad4bef65f.png">

I think we need to calculate the need for vertical independent of horizontal (instead of !horizontal).   I tried making a `isVertical:` method parallel to the exisitng `isHorizontal:`.   I changed:
```objc
#if TARGET_OS_OSX // [TODO(macOS ISS#2323203)
  BOOL isHorizontal = [self isHorizontal:_scrollView];
  [[self scrollView] setHasHorizontalScroller:isHorizontal];
  BOOL isVertical = [self isVertical:_scrollView];
  [[self scrollView] setHasVerticalScroller:isVertical];
#endif // ]TODO(macOS ISS#2323203)
```

and in `RCTScrollContentView.m` i changed to:
```objc
      Class horizontalScrollerClass = [scrollView isHorizontal:platformScrollView]
        ? [[platformScrollView horizontalScroller] class]
        : nil;
      Class verticalScrollerClass = [scrollView isVertical:platformScrollView]
        ? [[platformScrollView verticalScroller] class]
        : nil;
```

It's almost there.   The remaining issue is that in a `horizontal={true}` ScrollView, it still scrolls vertically, but without a visible scrollbar.   I can't figure out why:

![vertical-scroll](https://user-images.githubusercontent.com/30053638/84931568-cfb46100-b087-11ea-8fad-6498f45e2c63.gif)

Reply:
I played around with the view hierarchy debugger in Xcode and it looks like only `RCTScrollContentView` is resized. Its subviews are not. That got me thinking that we should probably fix this at the shadow layer, so I hacked together a thing just to confirm:

![image](https://user-images.githubusercontent.com/4123478/85031171-a5919a80-b17e-11ea-90e6-06622f1030f0.png)

Subviews are correctly resized and no longer clipped (green), but there are a couple of issues with background colors bleeding out (red). And since the shadow view doesn't have a reference to the real view, it doesn't know which scroller is currently visible. I'll continue looking for the correct place to make these adjustments but if you know exactly where to look, please let me know as I'm on unknown territory.

For reference, this is what I did:

```diff
diff --git a/React/Views/RCTShadowView.m b/React/Views/RCTShadowView.m
index ef74f2880..b8b2a0458 100644
--- a/React/Views/RCTShadowView.m
+++ b/React/Views/RCTShadowView.m
@@ -318,6 +318,12 @@ static void RCTProcessMetaPropsBorder(const YGValue metaProps[META_PROP_COUNT],
 }
 
 - (void)layoutSubviewsWithContext:(RCTLayoutContext)layoutContext
+{
+  [self layoutSubviewsWithContext:layoutContext adjustForLegacyScrollers:NO];
+}
+
+- (void)layoutSubviewsWithContext:(RCTLayoutContext)layoutContext
+         adjustForLegacyScrollers:(BOOL)adjustForLegacyScrollers
 {
   RCTLayoutMetrics layoutMetrics = self.layoutMetrics;
 
@@ -338,6 +344,23 @@ static void RCTProcessMetaPropsBorder(const YGValue metaProps[META_PROP_COUNT],
 
     RCTLayoutMetrics childLayoutMetrics = RCTLayoutMetricsFromYogaNode(childYogaNode);
 
+    if (adjustForLegacyScrollers) {
+      childLayoutMetrics.contentFrame.size =
+        [NSScrollView contentSizeForFrameSize:childLayoutMetrics.contentFrame.size
+                      horizontalScrollerClass:[NSScroller class]
+                        verticalScrollerClass:[NSScroller class]
+                                   borderType:NSNoBorder
+                                  controlSize:NSControlSizeRegular
+                                scrollerStyle:NSScrollerStyleLegacy];
+      childLayoutMetrics.frame.size =
+        [NSScrollView contentSizeForFrameSize:childLayoutMetrics.frame.size
+                      horizontalScrollerClass:[NSScroller class]
+                        verticalScrollerClass:[NSScroller class]
+                                   borderType:NSNoBorder
+                                  controlSize:NSControlSizeRegular
+                                scrollerStyle:NSScrollerStyleLegacy];
+    }
+
     layoutContext.absolutePosition.x += childLayoutMetrics.frame.origin.x;
     layoutContext.absolutePosition.y += childLayoutMetrics.frame.origin.y;
 
diff --git a/React/Views/ScrollView/RCTScrollContentShadowView.m b/React/Views/ScrollView/RCTScrollContentShadowView.m
index 16bf6147c..5bc3c06ae 100644
--- a/React/Views/ScrollView/RCTScrollContentShadowView.m
+++ b/React/Views/ScrollView/RCTScrollContentShadowView.m
@@ -11,11 +11,33 @@
 
 #import "RCTUtils.h"
 
+@interface RCTShadowView (RCTScrollContentShadowViewPrivate)
+- (void)layoutSubviewsWithContext:(RCTLayoutContext)layoutContext
+         adjustForLegacyScrollers:(BOOL)adjustForLegacyScrollers;
+@end
+
 @implementation RCTScrollContentShadowView
 
 - (void)layoutWithMetrics:(RCTLayoutMetrics)layoutMetrics
             layoutContext:(RCTLayoutContext)layoutContext
 {
+  if ([NSScroller preferredScrollerStyle] == NSScrollerStyleLegacy) {
+    layoutMetrics.contentFrame.size =
+      [NSScrollView contentSizeForFrameSize:layoutMetrics.contentFrame.size
+                    horizontalScrollerClass:[NSScroller class]
+                      verticalScrollerClass:[NSScroller class]
+                                 borderType:NSNoBorder
+                                controlSize:NSControlSizeRegular
+                              scrollerStyle:NSScrollerStyleLegacy];
+    layoutMetrics.frame.size =
+      [NSScrollView contentSizeForFrameSize:layoutMetrics.frame.size
+                    horizontalScrollerClass:[NSScroller class]
+                      verticalScrollerClass:[NSScroller class]
+                                 borderType:NSNoBorder
+                                controlSize:NSControlSizeRegular
+                              scrollerStyle:NSScrollerStyleLegacy];
+  }
+
   if (layoutMetrics.layoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
     // Motivation:
     // Yoga place `contentView` on the right side of `scrollView` when RTL layout is enforced.
@@ -29,4 +51,10 @@
   [super layoutWithMetrics:layoutMetrics layoutContext:layoutContext];
 }
 
+- (void)layoutSubviewsWithContext:(RCTLayoutContext)layoutContext
+{
+  [self layoutSubviewsWithContext:layoutContext
+         adjustForLegacyScrollers:[NSScroller preferredScrollerStyle] == NSScrollerStyleLegacy];
+}
+
 @end
diff --git a/React/Views/ScrollView/RCTScrollContentView.m b/React/Views/ScrollView/RCTScrollContentView.m
index e0e8d5311..9f4bfd793 100644
--- a/React/Views/ScrollView/RCTScrollContentView.m
+++ b/React/Views/ScrollView/RCTScrollContentView.m
@@ -21,25 +21,6 @@
 #else // [TODO(macOS ISS#2323203)
   // macOS also has a NSClipView in its hierarchy
   RCTScrollView *scrollView = (RCTScrollView *)self.superview.superview.superview;
-
-  if (scrollView != nil) {
-    // On macOS scroll indicators may float over the content view like they do in iOS
-    // or depending on system preferences they may be outside of the content view
-    // which means the clip view will be smaller than the scroll view itself.
-    // In such cases the content view layout must shrink accordingly otherwise
-    // the contents will overflow causing the scroll indicators to appear unnecessarily.
-    NSScrollView *platformScrollView = scrollView.scrollView;
-    if (platformScrollView.scrollerStyle == NSScrollerStyleLegacy) {
-      NSScroller *verticalScroller = platformScrollView.verticalScroller;
-      if (!verticalScroller.isHidden) {
-        frame.size.width -= verticalScroller.frame.size.width;
-      }
-      NSScroller *horizontalScroller = platformScrollView.horizontalScroller;
-      if (!horizontalScroller.isHidden) {
-        frame.size.height -= horizontalScroller.frame.size.height;
-      }
-    }
-  }
 #endif // ]TODO(macOS ISS#2323203)
 
   [super reactSetFrame:frame];
diff --git a/React/Views/ScrollView/RCTScrollView.m b/React/Views/ScrollView/RCTScrollView.m
index ad69c86e1..74d623436 100644
--- a/React/Views/ScrollView/RCTScrollView.m
+++ b/React/Views/ScrollView/RCTScrollView.m
@@ -187,7 +187,6 @@ RCT_NOT_IMPLEMENTED(- (instancetype)init)
     self.scrollEnabled = YES;
     self.hasHorizontalScroller = YES;
     self.hasVerticalScroller = YES;
-    self.autohidesScrollers = YES;
     self.panGestureRecognizer = [[NSPanGestureRecognizer alloc] initWithTarget:self action:@selector(handleCustomPan:)];
 #else // ]TODO(macOS ISS#2323203)
     [self.panGestureRecognizer addTarget:self action:@selector(handleCustomPan:)];
@@ -751,6 +750,11 @@ static inline void RCTApplyTransformationAccordingLayoutDirection(RCTPlatformVie
     [self react_updateClippedSubviewsWithClipRect:clipRect relativeToView:clipView];
     _lastClippedToRect = bounds;
   }
+
+#if TARGET_OS_OSX // [TODO(macOS ISS#2323203)
+  [[self scrollView] setHasHorizontalScroller:[self isHorizontal:_scrollView] || scrollsHorizontally];
+  [[self scrollView] setHasVerticalScroller:[self isVertical:_scrollView] || scrollsVertically];
+#endif // ]TODO(macOS ISS#2323203)
 }
 
 #if TARGET_OS_OSX // [TODO(macOS ISS#2323203)
```

Reply:
Latest iteration with local data for `RCTScrollContentShadowView`:
<img width="610" alt="image" src="https://user-images.githubusercontent.com/4123478/85109933-70cf2300-b212-11ea-9ed8-f49931a7971e.png">